3_computer_vision/ml_project_4_face_detection/backend/recognition.py

```diff
@@ -0,0 +1,165 @@
+import base64
+import io
+import json
+import pathlib
+from typing import Dict, List, Optional, Tuple
+
+import cv2
+import numpy as np
+import PIL
+import tensorflow as tf
+from google.cloud import aiplatform
+from keras.applications.resnet50 import ResNet50, preprocess_input
+from loguru import logger
+from pydantic import BaseModel, validator
+from tensorflow.keras.preprocessing.image import (
+    DirectoryIterator,
+    img_to_array,
+    load_img,
+)
+
+DEFAULT_IMAGE_LEN = 112
+
+DEFAULT_IMAGE_SIZE = (DEFAULT_IMAGE_LEN, DEFAULT_IMAGE_LEN)
+PATH_TO_CELEBRITY_MAPPING = (
+    pathlib.Path(__file__).resolve().parent / "idx_to_celebrity_name.json"
+)
+assert PATH_TO_CELEBRITY_MAPPING.exists(), f"Could not find {PATH_TO_CELEBRITY_MAPPING}"
+
+with open(PATH_TO_CELEBRITY_MAPPING, "r") as f:
+    loaded_idx_to_celebrity_name = json.load(f)
+
+
+def detect_celebrity(image: bytes, rekognition_client) -> Optional[str]:
+    response = rekognition_client.recognize_celebrities(Image={"Bytes": image})
+    logger.info(f"Response:\n{response}")
+
+    celebrities = response.get("CelebrityFaces")
+
+    return response["CelebrityFaces"][0]["Name"] if len(celebrities) > 0 else None
+
+
+def load_image_from_file(
+    image_path: pathlib.Path, target_image_size: Tuple[int, int] = DEFAULT_IMAGE_SIZE
+) -> PIL.Image:
+    return load_img(image_path, target_size=target_image_size)
+
+
+def load_local_model(model_path: pathlib.Path) -> Tuple[tf.keras.Model, Dict[int, str]]:
+    """
+    Returns a model saved locally in the SavedModel format, with a label
+    mapping dictionary that has been saved as an asset
+    """
+    saved_model = tf.keras.models.load_model(model_path)
+    with open(saved_model.asset.asset_path.numpy(), "r") as f:
+        loaded_idx_to_celebrity_name = json.load(f)
+
+    return saved_model, loaded_idx_to_celebrity_name
+
+
+def get_predicted_celebrity_name_from_idx(
+    data_streamer: DirectoryIterator, predicted_idx: int
+) -> str:
+    """
+    The mapping from celebrity name to index in the data streamer needs to be
+    inverted so that we can get the name of the predicted celebrity, given the
+    index
+    """
+    idx_to_celebrity_name = {v: k for k, v in data_streamer.class_indices.items()}
+    return str(idx_to_celebrity_name.get(predicted_idx))
+
+
+def make_prediction(
+    image: PIL.Image, model: tf.keras.Model, idx_to_celebrity_name: Dict[int, str]
+) -> Optional[str]:
+    logger.info(f"Input image: {image}")
+    example_image_array: np.ndarray = img_to_array(image)
+    logger.info(f"Image array shape: {example_image_array.shape}")
+    example_image_array = np.expand_dims(example_image_array, axis=0)
+    logger.info(f"After expanding dimensions: {example_image_array.shape}")
+    example_image_array = preprocess_input(example_image_array)
+    logger.info(f"After preprocessing: {example_image_array.shape}")
+
+    prediction_probabilities = model.predict(example_image_array)
```

Comment:
Something to consider for fun: you could also output the second and third predictions in addition to the argmax, and show that these can be used for further training (as negative examples). This helps improve convergence.